File: day11/part1.py
import copy

with open("input") as inputFile:
    input = inputFile.read().split("\n")[:-1]

    input = list(map(list, input))

changeIndexes = []

def numSurrounding(rowIndex, colIndex):
    count = 0

    for check in range(rowIndex - 1, rowIndex + 2): # Loop through the 3x3 grid of surrounding seats
        for seat in range(colIndex - 1, colIndex + 2):
            if (check < 0 or seat < 0 or check >= len(input) or seat >= len(input[check])): # Handles out-of-bounds as empty seats
                continue
            if (check == rowIndex and seat == colIndex): # Ignore the original seat
                continue
            if input[check][seat] == "#":
                count += 1
    return count

def prettyPrint(table):
    for i in list(map("".join, table)):
        print(i)

previousLayout = []

while (input != previousLayout):
    previousLayout = copy.deepcopy(input)
    for row in range(len(input)):
        for col in range(len(input[row])):
            if (input[row][col] == "L"): # Empty seat
                if (numSurrounding(row, col) == 0):
                    changeIndexes.append(["#", row, col])
            if (input[row][col] == "#"): # Full seat
                if (numSurrounding(row, col) >= 4):
                    changeIndexes.append(["L", row, col])

    for change in changeIndexes: # Make all the changes at once
        input[change[1]][change[2]] = change[0]
    changeIndexes.clear()

    # print("Shifted")

prettyPrint(input)

count = 0

for i in input:
    for j in i:
        if j == "#":
            count += 1

print(count)

File: day11/Part1.java
import java.util.ArrayList;
import java.util.Arrays;

class Part1 {

  private static int numSurrounding(int rowIndex, int colIndex, char[][] input) {

    int count = 0;

    for (int check = rowIndex - 1; check < rowIndex + 2; check++) {
      for (int seat = colIndex - 1; seat < colIndex + 2; seat++) {
        if (check < 0 || seat < 0 || check >= input.length || seat >= input[check].length || (check == rowIndex && seat == colIndex)) {
          continue;
        }
        if (input[check][seat] == '#') {
          count++;
        }
      }
    }

    return count;

  }

  public static void main(String[] args) {
    String[] inputData = Helpers.readFile("input");

    char[][] input = new char[inputData.length][inputData[0].length()];

    for (int cnt = 0; cnt < inputData.length; cnt++) {
      input[cnt] = inputData[cnt].toCharArray();
    }

    char[][] previousLayout = new char[input.length][input[0].length]; // Assumes that all rows will be of equal width

    ArrayList<String[]> changeIndexes = new ArrayList<>();

    while (!Helpers.deepEquals(input, previousLayout)) {
      previousLayout = Helpers.deepCopy(input);

      for (int row = 0; row < input.length; row++) {
        for (int col = 0; col < input[row].length; col++) {
          if (input[row][col] == 'L') {
            if (numSurrounding(row, col, input) == 0) {
              changeIndexes.add(new String[]{"#", String.valueOf(row), String.valueOf(col)});
            }
          } else if (input[row][col] == '#') {
            if (numSurrounding(row, col, input) >= 4) {
              changeIndexes.add(new String[]{"L", String.valueOf(row), String.valueOf(col)});
            }
          }
        }
      }

      for (String[] change : changeIndexes) {
        input[Integer.parseInt(change[1])][Integer.parseInt(change[2])] = change[0].toCharArray()[0];
      }
      changeIndexes.clear();

      // System.out.println("Shifted");

    }

    Helpers.prettyPrint(input);

    int count = 0;

    for (char[] i : input) {
      for (char j = 0; j < i.length; j++) {
        if (i[j] == '#') {
          count++;
        }
      }
    }

    System.out.println(count);

  }
}
